### PR TITLE
[INT-1162] Infinite scroll in file explorer

### PR DESCRIPTION
**Pull request by emmajsadams**
This PR implements infinite scroll pagination in our custom JuypterLab FileBrowser. It does so by setting up a scroll event listener that increments and decrements an index when user scrolls to the upper 20% and lower 20% respectively of the FileBrowser Contents DOM node. When this index changes the FileBrowser is re-rendered. Each index represents an offset of 200 Contents from the start of the overall cached Contents. Since only a 700 Contents are visible at any time regardless of index position or total Contents count this solution can scale with a large number of files (tested with 500k) since the low Contents limit of the Juypterlab FileBrowser is always respected. 

As part of testing this I also made a few changes to improve the performance of the each re-render of the FileBrowser.
* Cached the initial shallowResponse after navigating to a directory and reuse that in each call of _get. Making an GET request on each re-render is unnecessarily slow and is quite noticeable when infinite scroll causes a re-render. 
* Only re-render for infinite scroll not on each paginated request for Contents. These requests can happen safely in the background now without needing to update the pagination UI.

One feature currently not implemented is the ability to jump to anywhere in the Contents list. Since this solution works off an index this should be possible in a follow up PR. Dynamically changing the scrollbar height of the Contents DOM node to match the total count of cached Contents then determining where the user clicks to set the index should accomplish this. However this is tricky and worth splitting into a separate story.

### Demo
https://github.com/pachyderm/pachyderm/assets/401518/0ee94933-49be-461f-b5ec-cb144d9a2f06

### Why no library was used for infinite scroll
Most mature infinite scroll libraries (TanStack scroll, react-window, react-virtualized) require some sort of framework like React or Svelte and modify the DOM nodes as the user scrolls. This does not work with the Juypterlab FileBrowser code since even if we mounted a ReactWidget that output lumino styled `li` those li would not have the same event handlers setup by the FileBrowser. On top of that the FileBrowser breaks completely if Contents DOM nodes are modified outside the expected render loop.

### CODE REVIEW

File: jupyter-extension/src/plugins/mount/mountDrive.ts [hunk 1]
```diff
@@ -121,45 +133,41 @@ export class MountDrive implements Contents.IDrive {
...
 
+    const {start} = this.getContentsStart();
+    const {end} = this.getContentsEnd();
     return {
       ...shallowResponse,
-      content: contents.slice(
-        (this._model.page - 1) * MAX_NUM_CONTENTS_PAGE,
-        this._model.page * MAX_NUM_CONTENTS_PAGE,
-      ),
+      content: this._cache.filteredContents.slice(start, end),
```
Comment (emmajsadams) [inline jupyter-extension/src/plugins/mount/mountDrive.ts:170]: Hmm I noticed one issue with refreshing the page to a directory with a lot of files (as opposed to navigating to that directory from root/parent directory). Seems like it loads some sort of cache from the previous session which is then is overwrote once the first _get call finishes causing the scrollbar to reset to zero. 

It still works, but is a little odd from a UX perspective. The issue also existed before this PR. Still worth reviewing this as is since I can work on that issue and submit a separate PR for the fix (or follow up commit).

File: jupyter-extension/src/plugins/mount/mountDrive.ts [hunk 2]
```diff
@@ -2,16 +2,22 @@ import {ModelDB} from '@jupyterlab/observables';
...
 import {requestAPI} from '../../handler';
-import {Paging} from './paging';
 import {MOUNT_BROWSER_PREFIX} from './mount';
 
-// How many files to render in the FileBrowser UI per page
-const MAX_NUM_CONTENTS_PAGE = 100;
-// How many items to request per page
-const PAGINATION_NUMBER = 400;
+// How many Content li are visible at a given time for the user to scroll.
+const VISIBLE_CONTENT_LIS = 500;
```
Comment (bbonenfant) [inline jupyter-extension/src/plugins/mount/mountDrive.ts:12]: feel free to not do this, but maybe `VISIBLE_CONTENT_LI_COUNT` would be a clearer name?

File: jupyter-extension/src/plugins/mount/mountDrive.ts [hunk 3]
```diff
@@ -2,16 +2,22 @@ import {ModelDB} from '@jupyterlab/observables';
...
+// How many visible Content li that are padded on the top and bottom of the core set of visible Content li.
+// This padding enables the user to scroll while the Contents index changes.
+const VISIBLE_CONTENT_LI_PADDING = 200;
+// How many items to request per page.
+const PAGINATION_NUMBER = 1000;
+// How tall in pixels a Content li is.
+const LI_HEIGHT_PX = 25;
```
Comment (bbonenfant) [inline jupyter-extension/src/plugins/mount/mountDrive.ts:19]: Is this safe to set as a constant? This seems like something we don't really have control over -- could we extract this info from the browser at runtime?

File: jupyter-extension/src/plugins/mount/mountDrive.ts [hunk 4]
```diff
@@ -26,70 +32,76 @@ const DEFAULT_CONTENT_MODEL: Contents.IModel = {
...
   private _isDisposed = false;
+
+  // Signal emits `true` on navigating to a new directory and `false` on response.
+  private _loading = new Signal<this, boolean>(this);
+  // Contents cache for the Drive used to avoid the small Contents limit that can be visible in a FileBrowser.
   private _cache: {
     key: string | null;
     now: number | null;
     contents: Contents.IModel[];
+    filteredContents: Contents.IModel[];
```
Comment (bbonenfant) [inline jupyter-extension/src/plugins/mount/mountDrive.ts:49]: Just for clarification, why this addition to the cache?

**Review by bbonenfant — APPROVED** [review]
This looks good to me, but the portions were a bit outside my expertise. If you'd like more substantial feed back we might be able to convince someone from console to take a look?

File: jupyter-extension/src/plugins/mount/mountDrive.ts (hunk 3 above)
Comment (emmajsadams) [reply jupyter-extension/src/plugins/mount/mountDrive.ts:19]: Good point! I can grab the clientHeight from the first contentsNode.childNodes[0] instead of hard coding this.

File: jupyter-extension/src/plugins/mount/mountDrive.ts (hunk 2 above)
Comment (emmajsadams) [reply jupyter-extension/src/plugins/mount/mountDrive.ts:12]: I like that name better will change!

File: jupyter-extension/src/plugins/mount/mountDrive.ts (hunk 1 above)
Comment (emmajsadams) [reply jupyter-extension/src/plugins/mount/mountDrive.ts:170]: Fixed this issue in the next commit I am pushing!

File: jupyter-extension/src/plugins/mount/mountDrive.ts (hunk 4 above)
Comment (emmajsadams) [reply jupyter-extension/src/plugins/mount/mountDrive.ts:49]: `_cache.contents` refers to all the files for the current directory whereas `_cache.filteredContents` refers to the contents currently filtered by the value the user has typed into the FilterBox above the FileBrowser.

Since `_get` is called each time the user changes the `_index` by scrolling, we want to avoid filtering the cached contents each call since the cached contents could be huge. Thus the filtering only occurs when the user changes directory or changes the filtering then is cached.

Comment (bbonenfant) [reply jupyter-extension/src/plugins/mount/mountDrive.ts:47]: Ahh ok, the piece I was missing was that _get is called for every scroll. thanks!

File: jupyter-extension/cypress/integration/mount.ts [hunk 5]
```diff
@@ -18,7 +18,9 @@ describe('mount', () => {
     cy.findAllByText('/ pfs').should('have.length', 2);
     cy.findAllByText('Load').first().click();
     cy.findAllByText('Unload').should('have.length', 1);
-    cy.findAllByText('default_images').first().click();
+    cy.get('#jupyterlab-pachyderm-browser-pfs')
```
Comment (emmajsadams) [inline jupyter-extension/cypress/integration/mount.ts:21]: Not exactly sure why this change is needed now, but the problem with searching `findByText` without an id is that `default_images` also appears as text in the repo list. When I debugged this locally the text is the element that was clicked which obviously does nothing. Changing this to findByText in the FileBrowser ID scope fixes this issue.

File: jupyter-extension/cypress/integration/repo-datum-mode-switch.ts [hunk 6]
```diff
@@ -30,9 +30,14 @@ describe('switching between repo and datum mode', () => {
 
     cy.findAllByText('Unload').should('have.length', 2);
     cy.wait(3000);
-    cy.findAllByText('default_images').first().click();
+    cy.get('#jupyterlab-pachyderm-browser-pfs')
+      .findByText('default_images')
+      .dblclick();
     cy.findAllByText('liberty.png').should('have.length', 1);
-    cy.findAllByText('default_images_branch').first().click();
+    cy.get('#jupyterlab-pachyderm-browser-pfs').findByText('/ pfs').click();
```
Comment (emmajsadams) [inline jupyter-extension/cypress/integration/repo-datum-mode-switch.ts:37]: This was really strange I honestly do not know how this test passed before this change. What this test is doing is clicking on a mounted repo folder, checking the file is there, then repeating that for a second mounted repo folder. The problem with the test is that you have to navigate back to the root directory to see the second mounted repo folder, but the test does not do that. Thus I added one step that clicks on the `/ pfs` breadcrumb to make it pass.

**Comment by emmajsadams** [comment]
Ok I responded to your comments and included two new changes @bbonenfant.

* Disabled FileBrowser restore so that the user scrolling immediately isn't interrupted.
* Cached the initial shallowResponse after navigating to a directory and reuse that in each call of _get. Making an GET request on each re-render is unnecessarily slow and is quite noticeable when infinite scroll causes a re-render. 

I commented on the changes that fix the tests to explain why they were needed. I did figure out the issue with running Cypress tests locally, but it requires us migrating to the latest cypress. I will split that out into a separate PR since it results in a bunch of changes.

As far as having console approval maybe lets wait until the next PR which allows the user to jump to anywhere in the Contents list via the scroll bar https://pachyderm.atlassian.net/browse/INT-1203?

**Comment by bbonenfant** [comment]
Those cypress changes are weird. Hopefully moving to a more recent version will make these tests easier to work with. LGTM -- merge when you're ready
